File: epi_judge_java/epi/LowestCommonAncestorWithParent.java
package epi;

import epi.test_framework.*;

public class LowestCommonAncestorWithParent {

    public static BinaryTree<Integer> lca(BinaryTree<Integer> n1, BinaryTree<Integer> n2) {
        int d1 = 0;
        var runner = n1;
        while (runner != null) {
            d1++;
            runner = runner.parent;
        }
        int d2 = 0;
        runner = n2;
        while (runner != null) {
            d2++;
            runner = runner.parent;
        }
        while (d1 > d2) {
            d1--;
            n1 = n1.parent;
        }
        while (d2 > d1) {
            d2--;
            n2 = n2.parent;
        }
        while (n1 != n2) {
            n1 = n1.parent;
            n2 = n2.parent;
        }
        return n1;
    }

    @EpiTest(testDataFile = "lowest_common_ancestor.tsv")
    public static int lcaWrapper(TimedExecutor executor, BinaryTree<Integer> tree,
                                 Integer key0, Integer key1) throws Exception {
        BinaryTree<Integer> node0 = BinaryTreeUtils.mustFindNode(tree, key0);
        BinaryTree<Integer> node1 = BinaryTreeUtils.mustFindNode(tree, key1);

        BinaryTree<Integer> result = executor.run(() -> lca(node0, node1));

        if (result == null) {
            throw new TestFailure("Result can not be null");
        }
        return result.data;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "LowestCommonAncestorWithParent.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/lowest_common_ancestor_with_parent.cc
#include "binary_tree_with_parent_prototype.h"
#include "test_framework/binary_tree_utils.h"
#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"

BinaryTreeNode<int> *Lca(const unique_ptr<BinaryTreeNode<int>> &node1,
                         const unique_ptr<BinaryTreeNode<int>> &node2) {

    auto r1 = node1.get();
    int h1 = 0;
    while (r1 != nullptr) {
        h1++;
        r1 = r1->parent;
    }
    auto r2 = node2.get();
    int h2 = 0;
    while (r2 != nullptr) {
        h2++;
        r2 = r2->parent;
    }

    r1 = node1.get();
    r2 = node2.get();
    int diff = h1 - h2;
    if (diff > 0) {
        while (diff--) {
            r1 = r1->parent;
        }
    } else if (diff < 0) {
        diff = -diff;
        while (diff--) {
            r2 = r2->parent;
        }
    }

    while (r1 != r2) {
        r1 = r1->parent;
        r2 = r2->parent;
    }
    return r1;
}

int LcaWrapper(TimedExecutor &executor,
               const unique_ptr<BinaryTreeNode<int>> &tree, int key0,
               int key1) {
    const unique_ptr<BinaryTreeNode<int>> &node0 = MustFindNode(tree, key0);
    const unique_ptr<BinaryTreeNode<int>> &node1 = MustFindNode(tree, key1);

    auto result = executor.Run([&] { return Lca(node0, node1); });

    if (!result) {
        throw TestFailure("Result can not be nullptr");
    }
    return result->data;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"executor", "tree", "key0", "key1"};
    return GenericTestMain(args, "lowest_common_ancestor_with_parent.cc",
                           "lowest_common_ancestor.tsv", &LcaWrapper,
                           DefaultComparator{}, param_names);
}
